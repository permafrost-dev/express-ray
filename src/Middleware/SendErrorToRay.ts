import { ray } from 'node-ray';

export const sendErrorToRay = (err, req, res, next, rayInstance: any = null) => {
    const { method, url, params, query } = req;

    rayInstance = rayInstance ?? ray();

    rayInstance.table(
        {
            error: err.message,
            method,
            url,
            params,
            query,
        },
        'Error',
    );

    if (res.headersSent) {
        return next(err);
    }

    next(err);
};

export default sendErrorToRay;
