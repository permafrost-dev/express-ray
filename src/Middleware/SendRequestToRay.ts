import { ray } from 'node-ray';
import micromatch from 'micromatch';
export interface SendRequestToRayOptions {
    paths?: {
        include?: string[];
        ignore?: string[];
    };
}

export const sendRequestToRay = (options: SendRequestToRayOptions = {}) => {
    const matchPaths = options.paths?.include || ['*'];
    const ignorePaths = options.paths?.ignore || [];

    const result = (req, res, next, rayInstance: any = null) => {
        const matchedPath = micromatch.isMatch(req.path, matchPaths);
        const isIgnoredPath = micromatch.isMatch(req.path, ignorePaths);

        if (!matchedPath || isIgnoredPath) {
            return next();
        }

        const { method, url, params, query, headers } = req;

        rayInstance = rayInstance ?? ray();

        const data = {
            method,
            url,
            params,
            query,
            headers,
        };

        if (method.toLowerCase() === 'post') {
            data['body'] = req.body;
        }

        rayInstance.table(data, 'Request');

        next();
    };

    return result;
};

export default sendRequestToRay;
