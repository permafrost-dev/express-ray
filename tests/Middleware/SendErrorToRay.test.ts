/* eslint-disable @typescript-eslint/no-empty-function */

import { sendErrorToRay } from '@/Middleware/SendErrorToRay';
import { FakeRay } from '@tests/TestClasses/FakeRay';

it('sends an error to ray', () => {
    const rayInstance = new FakeRay();

    const request = {
        method: 'GET',
        url: '/',
        params: {},
        query: {},
    };

    sendErrorToRay(new Error('test'), request, {}, () => {}, rayInstance);

    expect(rayInstance.sentPayloads).toHaveLength(1);
    expect(rayInstance.sentPayloads).toMatchSnapshot();
});
