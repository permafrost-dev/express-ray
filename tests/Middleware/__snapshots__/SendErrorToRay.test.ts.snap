// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sends an error to ray 1`] = `
Array [
  Object {
    "data": Object {
      "error": "test",
      "method": "GET",
      "params": Object {},
      "query": Object {},
      "url": "/",
    },
    "label": "Error",
  },
]
`;
